persistence/sql/persister_grant_jwk.go

```diff
@@ -0,0 +1,196 @@
+		return trust.Grant{}, sqlcon.HandleError(err)
+	}
+
+	return p.jwtGrantFromSQlData(data), nil
+}
+
+func (p *Persister) DeleteGrant(ctx context.Context, id string) error {
+	grant, err := p.GetConcreteGrant(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	return p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
+		if err := p.Connection(ctx).Destroy(&trust.SQLData{ID: grant.ID}); err != nil {
+			return sqlcon.HandleError(err)
+		}
+
+		return p.DeleteKey(ctx, grant.PublicKey.Set, grant.PublicKey.KeyID)
+	})
+}
+
+func (p *Persister) GetGrants(ctx context.Context, limit, offset int, optionalIssuer string) ([]trust.Grant, error) {
+	grantsData := make([]trust.SQLData, 0)
+
+	query := p.Connection(ctx).Paginate(offset/limit+1, limit).Order("id")
+	if optionalIssuer != "" {
+		query = query.Where("issuer = ?", optionalIssuer)
+	}
+
+	if err := query.All(&grantsData); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	grants := make([]trust.Grant, 0, len(grantsData))
+	for _, data := range grantsData {
+		grants = append(grants, p.jwtGrantFromSQlData(data))
+	}
+
+	return grants, nil
+}
+
+func (p *Persister) CountGrants(ctx context.Context) (int, error) {
+	n, err := p.Connection(ctx).Count(&trust.SQLData{})
+	return n, sqlcon.HandleError(err)
+}
+
+func (p *Persister) GetPublicKey(ctx context.Context, issuer string, subject string, keyId string) (*jose.JSONWebKey, error) {
+	var data trust.SQLData
+	query := p.Connection(ctx).
+		Where("issuer = ?", issuer).
+		Where("subject = ?", subject).
+		Where("key_id = ?", keyId)
+	if err := query.First(&data); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	keySet, err := p.GetKey(ctx, data.KeySet, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &keySet.Keys[0], nil
+}
+
+func (p *Persister) GetPublicKeys(ctx context.Context, issuer string, subject string) (*jose.JSONWebKeySet, error) {
+	grantsData := make([]trust.SQLData, 0)
+	query := p.Connection(ctx).
+		Where("issuer = ?", issuer).
+		Where("subject = ?", subject)
+	if err := query.All(&grantsData); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	if len(grantsData) == 0 {
+		return &jose.JSONWebKeySet{}, nil
+	}
+
+	// because keys must be grouped by issuer, we can retrieve set name from first grant
+	keySet, err := p.GetKeySet(ctx, grantsData[0].KeySet)
+	if err != nil {
+		return nil, err
+	}
+
+	// find keys, that belong to grants
+	filteredKeySet := &jose.JSONWebKeySet{}
+	for _, data := range grantsData {
+		if keys := keySet.Key(data.KeyID); len(keys) > 0 {
+			filteredKeySet.Keys = append(filteredKeySet.Keys, keys...)
+		}
+	}
+
+	return filteredKeySet, nil
+}
+
+func (p *Persister) GetPublicKeyScopes(ctx context.Context, issuer string, subject string, keyId string) ([]string, error) {
+	var data trust.SQLData
+	query := p.Connection(ctx).
+		Where("issuer = ?", issuer).
+		Where("subject = ?", subject).
+		Where("key_id = ?", keyId)
```

Comment:
Well in all databases used here, unique constrain adds index automatically.

Reply:
> Well in all databases used here, unique constrain adds index automatically.

You're right! I'm not familiar with the others but cockroach docs say the following: 
```
Columns with the UNIQUE constraint automatically have an index created with the name <table name>_<columns>_key. To avoid having two identical indexes, you should not create indexes that exactly match the UNIQUE constraint's columns and order.
```